fix: avoid iterating over a changing table in `run_threads`

Guldoman opened the pull request:
This is done to avoid iterating over a table that can change in the meantime. More precisely the issue appears if a thread is removed from the table, we yield early because we reached the end of the frame, and a new thread is added before the next iteration.

For example:
```lua
local lost_time = false
core.add_thread(function()
  -- force early yield
  local t0 = system.get_time()
  while system.get_time() - t0 < .1 do end
  lost_time = true
end, "a")

local step = core.step
function core.step()
  if lost_time then
    -- add a new thread while run_threads hasn't finished iterating
    core.add_thread(function()end, "a1")
    lost_time = false
  end
  return step()
end
```
would crash with `invalid key to 'next'`.

takase1121 commented:
Will it be more efficient to store thread addition/removal in another table and retroactively apply it? I'm not sure, but then it's unlikely to have a lot of threads where this is going to matter at all.

Guldoman commented:
> Will it be more efficient to store thread addition/removal in another table and retroactively apply it?

Main problem is that if anything else externally removes entries from the table, we might get the same issue.

This makes me think: what should we do if something externally removes the entry? Should we execute it anyways one last time, or should we check if it still exists before executing it?

takase1121 commented:
> Main problem is that if anything else externally removes entries from the table, we might get the same issue.

This is just an incredibly stupid thing to do. I don't think there is a way to catch this, but I don't think this scenario should be supported.

Aside from that, how does weak key tables interact with `pair()`? Will the GC kill the table mid iteration?

Guldoman commented:
With the last commit I fixed an issue that caused skipped entries (an issue even before this PR: after an entry was removed, it skipped the next), and deleted the wrong ones (only an issue with this PR).

takase1121 reviewed (approved):
Let's get this done first.
